controllers/hostedcontrolplane/hostedcontrolplane.go

```diff
@@ -329,3 +381,288 @@ func (r *HostedControlPlaneReconciler) SetupWithManager(mgr ctrl.Manager) error
+}
+
+func (APIClient *APIClient) getDynatraceHTTPMonitorID(ctx context.Context, log logr.Logger, hostedcontrolplane *v1beta1.HostedControlPlane) (string, error) {
+	labels := hostedcontrolplane.GetLabels()
+	dynatraceHttpMonitorId, exists := labels[httpMonitorLabel]
+	if exists {
+		return dynatraceHttpMonitorId, nil
+	}
+	return "", fmt.Errorf("key '%s' not found in labels", httpMonitorLabel)
+}
+
+func (r *HostedControlPlaneReconciler) UpdateHostedControlPlaneLabels(ctx context.Context, hostedcontrolplane *v1beta1.HostedControlPlane, key, value string) error {
+	labels := hostedcontrolplane.GetLabels()
+	labels[key] = value
+	hostedcontrolplane.SetLabels(labels)
+
+	err := r.Client.Update(ctx, hostedcontrolplane)
+	if err != nil {
+		return fmt.Errorf("error updating hostedcontrolplane monitor: %v", err)
+	}
+	return nil
+}
+
+func (r *HostedControlPlaneReconciler) GetAPIServerHostname(hostedcontrolplane *v1beta1.HostedControlPlane, log logr.Logger) (string, error) {
+	for _, service := range hostedcontrolplane.Spec.Services {
+		if service.Service == "APIServer" {
+			return service.ServicePublishingStrategy.Route.Hostname, nil
+		}
+	}
+	return "", fmt.Errorf("APIServer service not found in the hostedcontrolplane")
+}
+
+func (APIClient *APIClient) getDynatraceEquivalentClusterRegionId(hostedcontrolplane *v1beta1.HostedControlPlane) (string, error) {
+	openShiftToAwsRegions := map[string]string{
+		"us": "N. Virginia",
+		"af": "Cape Town",
+		"ap": "Seoul",
+		"ca": "Montreal",
+		"eu": "Frankfurt",
+		"me": "Bahrain",
+		"sa": "São Paulo",
+	}
+	clusterRegion := hostedcontrolplane.Spec.Platform.AWS.Region
+	regionCode := strings.Split(clusterRegion, "-")[0]
+	// Look up the location name based on the region code in map
+	locationName, ok := openShiftToAwsRegions[regionCode]
+	if !ok {
+		return "", fmt.Errorf("location not found for region code: %s and region: %s", regionCode, clusterRegion)
+	}
+
+	resp, err := APIClient.makeRequest("GET", "/synthetic/locations", nil)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to fetch locations. Status code: %d", resp.StatusCode)
+	}
+
+	//return location id from response body
+	var data map[string][]map[string]interface{}
+	err = json.NewDecoder(resp.Body).Decode(&data)
+	if err != nil {
+		return "", err
+	}
+	locations, ok := data["locations"]
+	if !ok {
+		return "", fmt.Errorf("locations not found in response")
+	}
+	for _, loc := range locations {
+		if loc["name"] == locationName && loc["type"] == "PUBLIC" && loc["cloudPlatform"] == "AMAZON_EC2" {
+			return loc["entityId"].(string), nil
+		}
+	}
+
+	return "", fmt.Errorf("location '%s' not found", locationName)
+}
+
+func (APIClient *APIClient) createDynatraceHTTPMonitor(monitorName, apiUrl, clusterId, dynatraceEquivalentClusterRegionId string) (string, error) {
+	monitorConfig := map[string]interface{}{
```

Comment:
Seeing all of these `map[string]interface{}` definitions is a smell to me.

Would it maybe make more sense to use this as a [template](https://pkg.go.dev/text/template) if we need some of the more advanced functions?  Or could we define this as its own struct if we need to interact with it on a more granular level?

Reply:
We are using template and structs now. Thanks!